upgrader_fsm.go

```diff
@@ -0,0 +1,86 @@
+package tableroll
+
+import "fmt"
+
+// upgraderState represents a small finite state machine. It has the following transitions:
+// ∅                     → CheckingOwner
+// CheckingOwnership     → AwaitingOwnership
+// CheckingOwnership     → Owner
+// AwaitingOwnership     → Owner
+// Owner                 → TransferringOwnership
+// TransferringOwnership → Owner
+// TransferringOwnership → Draining
+//
+// The meaning of each state is described above the state's definition below.
+type upgraderState string
+
+const (
+	// CheckingOwnership is the initial state. It indicates this upgrader is
+	// trying to connect to the current owner to determine if they exist.
+	upgraderStateCheckingOwner upgraderState = "checking-owner"
+	// AwaitingOwnership is transitioned to if a current owner is detected. In
+	// this state, the upgrader is awaiting the new owner to pass file
+	// descriptors.
+	upgraderStateAwaitingOwnership upgraderState = "awaiting-ownership"
+	// Owner is the state of an upgrader that has successfully either upgraded or
+	// determined that it is the sole process and should thus take ownership.
+	upgraderStateOwner = "owner"
+	// TransferringOwnership is the state of an upgrader that has received a
+	// request from a new process to pass over its FDs, but either has not passed
+	// them all over, or has not yet received a ready.
+	upgraderStateTransferringOwnership = "transferring-ownership"
+	// Draining is the state a process is in after a new owner has taken over.
+	upgraderStateDraining = "draining"
+	// Stopped is the state a process is in after it has completed draining or
+	// has been marked to stop.
+	upgraderStateStopped = "stopped"
+)
+
+func (u *upgraderState) canTransitionTo(state upgraderState) error {
+	validTransitions := map[upgraderState][]upgraderState{
+		upgraderStateCheckingOwner: []upgraderState{
+			upgraderStateAwaitingOwnership,
+			upgraderStateOwner,
+			upgraderStateStopped,
+		},
+		upgraderStateAwaitingOwnership: []upgraderState{
+			upgraderStateOwner,
+			upgraderStateStopped,
+		},
```

Comment:
it appears that the awaitingownership state is never used anywhere. maybe it can be removed? (or there's a logic bug elsewhere?)

Reply:
Good catch, originally that seemed useful to model, but ultimately all state is just serially handled in the constructor up to the `serveUpgrade` bit, so that ended up being the main thing that mattered.

Nixed that state, the startup procedure is just serialized naturally so it's unneeded.